fix(sql): fix comparing Date columns with date literals

jerrinot opened the pull request:
Consider this query: `select * from tab where date = '1970-01-01'`

Without these new functions QuestDB resorts to picking the overloaded equal function which accepts 2 longs. Why?
1. There is no exact match for eq(Date, Date) - such function does not exist at all!
2. The date column is `DateColumn` and the Date type can be casted to Long.
   See the overloading priority: /* 7  DATE      */, {DATE, TIMESTAMP, LONG}
3. The second parameter is `StrConstant` and String can be also casted to Long.
   Again, the overloading priority:  /* 11 STRING    */, {STRING, CHAR, DOUBLE, LONG, INT, FLOAT, SHORT, BYTE}
4. Hence, the function parser picks eq(Long, Long).

But then the `StrFunction.getLong()` fails to parse '1970-01-01' as a long. It expects a simple number. It does not know it's supposed to treat the argument as a Date.

The same query works if you change the column type from Date to Timestamp. Why? Timestamp does not rely on casting - it implements Timestamp-specific functions where necessary. So I copied this for Date.

ideoma commented:
#### [PR Coverage check]

:heart_eyes: **pass** : 42 / 59 (71.19%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/FunctionParser.java|4|13|30.77%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/lt/LtDateFunctionFactory.java|14|22|63.64%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastTimestampToDateFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastFloatToDateFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastStrToDateFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/eq/EqDateFunctionFactory.java|7|7|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/AbstractCastToDateFunction.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastByteToDateFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastBooleanToDateFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastLongToDateFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastIntToDateFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastSymbolToDateFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/AbstractUnaryDateFunction.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastCharToDateFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastDoubleToDateFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastShortToDateFunctionFactory.java|1|1|100.00%|

ideoma commented:
#### [PR Coverage check]

:heart_eyes: **pass** : 42 / 59 (71.19%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/FunctionParser.java|4|13|30.77%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/lt/LtDateFunctionFactory.java|14|22|63.64%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastTimestampToDateFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastFloatToDateFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastStrToDateFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/eq/EqDateFunctionFactory.java|7|7|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/AbstractCastToDateFunction.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastByteToDateFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastBooleanToDateFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastLongToDateFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastIntToDateFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastSymbolToDateFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/AbstractUnaryDateFunction.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastCharToDateFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastDoubleToDateFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastShortToDateFunctionFactory.java|1|1|100.00%|

ideoma commented:
#### [PR Coverage check]

:heart_eyes: **pass** : 61 / 73 (83.56%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/functions/lt/LtDateFunctionFactory.java|14|22|63.64%|
|:large_blue_circle:|io/questdb/griffin/FunctionParser.java|9|13|69.23%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastTimestampToDateFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastFloatToDateFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastStrToDateFunctionFactory.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/eq/EqDateFunctionFactory.java|7|7|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/AbstractCastToDateFunction.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastByteToDateFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastBooleanToDateFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastLongToDateFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastIntToDateFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastSymbolToDateFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/AbstractUnaryDateFunction.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastCharToDateFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastDoubleToDateFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/datetime/millitime/DateFormatUtils.java|10|10|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastShortToDateFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlUtil.java|3|3|100.00%|

ideoma commented:
#### [PR Coverage check]

:heart_eyes: **pass** : 69 / 73 (94.52%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/FunctionParser.java|9|13|69.23%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastTimestampToDateFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastFloatToDateFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastStrToDateFunctionFactory.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/eq/EqDateFunctionFactory.java|7|7|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/AbstractCastToDateFunction.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastByteToDateFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastBooleanToDateFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastLongToDateFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastIntToDateFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/lt/LtDateFunctionFactory.java|22|22|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastSymbolToDateFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/AbstractUnaryDateFunction.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastCharToDateFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastDoubleToDateFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/datetime/millitime/DateFormatUtils.java|10|10|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastShortToDateFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlUtil.java|3|3|100.00%|

ideoma commented:
#### [PR Coverage check]

:heart_eyes: **pass** : 72 / 74 (97.30%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/FunctionParser.java|12|14|85.71%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastTimestampToDateFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastFloatToDateFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastStrToDateFunctionFactory.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/eq/EqDateFunctionFactory.java|7|7|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/AbstractCastToDateFunction.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastByteToDateFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastBooleanToDateFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastLongToDateFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastIntToDateFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/lt/LtDateFunctionFactory.java|22|22|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastSymbolToDateFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/AbstractUnaryDateFunction.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastCharToDateFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastDoubleToDateFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/datetime/millitime/DateFormatUtils.java|10|10|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastShortToDateFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlUtil.java|3|3|100.00%|

bluestreak01 commented:
@jerrinot this is a solid PR! I do think tho that title should be a "fix" or "feat". It is user facing problem. Also description needs a succint TLDR

jerrinot commented:
@bluestreak01 thanks. I think "fix" is the more appropriate categorization. As most users will perceive the current behaviour as buggy.

bziobrowski commented on core/src/main/java/io/questdb/griffin/FunctionParser.java:
```diff
@@ -781,26 +785,31 @@ private Function createFunction(
                 }
             } else if ((argTypeTag == ColumnType.STRING || argTypeTag == ColumnType.SYMBOL) && sigArgTypeTag == ColumnType.TIMESTAMP) {
                 // convert arguments if necessary
```
Why not push `assert arg.isConstant()` to if condition here and below  ? 
It could `explode` with AssertionError  if assertions are enabled.

jerrinot replied:
That's actually something I wanted to ask you: Is it possible that arg is ever not a constant? 

This is my understanding: The overloading rules as defined in `ColumnType` do not allow matching a function declared as receiving `Date` with a String argument. So a function which declares an argument of a type `Date` will match with an actual String argument if and only if the string is a constant. Only because of the code/hack at the lines 679-682. An that applies to constants only. No other case is allowed. And the assertion is meant to communicate this invariant. Does it make sense?

bziobrowski replied:
You're right, non-const case is not reachable, so I guess it's fine .

bluestreak01 commented:
@jerrinot is this draft draft or ready to go draft?

jerrinot commented:
@bluestreak01: I would like to add some more tests, I believe @marregui also wanted to add some tests. 
+ I want to try the thing we discussed the other day: Do not modify FunctionParser, instead implement comparison functions which accept combinations of String and Date arguments.

marregui commented:
hi @bluestreak01 package io.questdb.griffin.engine.functions.cast is a led zepelin. In the long term we would benefit from generating such classes on compile time, the full extent, or found an alternative approach. I will add a few tests, and will double check the approach, but if it works well, would you agree that it should be enough?

ideoma commented:
#### [PR Coverage check]

:heart_eyes: **pass** : 153 / 155 (98.71%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/FunctionParser.java|38|40|95.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastTimestampToDateFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/geohash/GeoHashFromCoordinatesFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastFloatToDateFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastStrToDateFunctionFactory.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastStrToGeoHashFunctionFactory.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/eq/EqDateFunctionFactory.java|7|7|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/AbstractCastToDateFunction.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/LongNullUtils.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastByteToDateFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/conditional/CaseCommon.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/rnd/RndGeoHashFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/ColumnType.java|31|31|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bool/InTimestampTimestampFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/WhereClauseParser.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/GeoHashes.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/ExpressionParser.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/constants/GeoHashTypeConstant.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastBooleanToDateFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/constants/Constants.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastLongToDateFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastIntToDateFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/catalogue/TypeOfFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/lt/LtDateFunctionFactory.java|22|22|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastSymbolToDateFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/GeoHashUtil.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlException.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/AbstractUnaryDateFunction.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastCharToDateFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastDoubleToDateFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/text/types/GeoHashAdapter.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/datetime/millitime/DateFormatUtils.java|10|10|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastShortToDateFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlUtil.java|3|3|100.00%|

ideoma commented:
#### [PR Coverage check]

:heart_eyes: **pass** : 153 / 155 (98.71%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/FunctionParser.java|38|40|95.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastTimestampToDateFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/geohash/GeoHashFromCoordinatesFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastFloatToDateFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastStrToDateFunctionFactory.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastStrToGeoHashFunctionFactory.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/eq/EqDateFunctionFactory.java|7|7|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/AbstractCastToDateFunction.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/LongNullUtils.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastByteToDateFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/conditional/CaseCommon.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/rnd/RndGeoHashFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/ColumnType.java|31|31|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bool/InTimestampTimestampFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/WhereClauseParser.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/GeoHashes.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/ExpressionParser.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/constants/GeoHashTypeConstant.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastBooleanToDateFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/constants/Constants.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastLongToDateFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastIntToDateFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/catalogue/TypeOfFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/lt/LtDateFunctionFactory.java|22|22|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastSymbolToDateFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/GeoHashUtil.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlException.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/AbstractUnaryDateFunction.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastCharToDateFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastDoubleToDateFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/text/types/GeoHashAdapter.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/datetime/millitime/DateFormatUtils.java|10|10|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastShortToDateFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlUtil.java|3|3|100.00%|

jerrinot commented:
@bluestreak01: I tried the experiment we discussed the other day: I introduced `EqDateStringFunctionFactory` with this signature: `=(Ms)` and reverted the functional changes inside `FunctionParser`. However, this had unexpected side effects. 

Consider this table: `create table tab (ts timestamp);`  and this SQL: `select * from tab where ts = '2022-01-01'` This is supposed to match the `EqTimestampFunctionFactory`. However, it started to match the `EqDateStringFunctionFactory` instead! Because the 2nd argument matched exactly (string constant) and the `Timestamp` column can be cast to `Date`. This is not undesirable, we do not want to cast Timestamp to a coarse-grained Date for no good reason. Thus I think the tiny change in `FunctionParser` is safer.

ideoma commented:
#### [PR Coverage check]

:heart_eyes: **pass** : 153 / 155 (98.71%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/FunctionParser.java|38|40|95.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastTimestampToDateFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/geohash/GeoHashFromCoordinatesFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastFloatToDateFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastStrToDateFunctionFactory.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastStrToGeoHashFunctionFactory.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/eq/EqDateFunctionFactory.java|7|7|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/AbstractCastToDateFunction.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/LongNullUtils.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastByteToDateFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/conditional/CaseCommon.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/rnd/RndGeoHashFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/ColumnType.java|31|31|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bool/InTimestampTimestampFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/WhereClauseParser.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/GeoHashes.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/ExpressionParser.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/constants/GeoHashTypeConstant.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastBooleanToDateFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/constants/Constants.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastLongToDateFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastIntToDateFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/catalogue/TypeOfFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/lt/LtDateFunctionFactory.java|22|22|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastSymbolToDateFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/GeoHashUtil.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlException.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/AbstractUnaryDateFunction.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastCharToDateFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastDoubleToDateFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/text/types/GeoHashAdapter.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/datetime/millitime/DateFormatUtils.java|10|10|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastShortToDateFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlUtil.java|3|3|100.00%|

ideoma commented:
#### [PR Coverage check]

:heart_eyes: **pass** : 153 / 155 (98.71%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/FunctionParser.java|38|40|95.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastTimestampToDateFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/geohash/GeoHashFromCoordinatesFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastFloatToDateFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastStrToDateFunctionFactory.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastStrToGeoHashFunctionFactory.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/eq/EqDateFunctionFactory.java|7|7|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/AbstractCastToDateFunction.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/LongNullUtils.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastByteToDateFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/conditional/CaseCommon.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/rnd/RndGeoHashFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/ColumnType.java|31|31|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bool/InTimestampTimestampFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/WhereClauseParser.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/GeoHashes.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/ExpressionParser.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/constants/GeoHashTypeConstant.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastBooleanToDateFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/constants/Constants.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastLongToDateFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastIntToDateFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/catalogue/TypeOfFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/lt/LtDateFunctionFactory.java|22|22|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastSymbolToDateFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/GeoHashUtil.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlException.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/AbstractUnaryDateFunction.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastCharToDateFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastDoubleToDateFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/text/types/GeoHashAdapter.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/datetime/millitime/DateFormatUtils.java|10|10|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastShortToDateFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlUtil.java|3|3|100.00%|
